providers/common/ciphers/aes_gcm_s390x.inc

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright 2001-2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * IBM S390X support
+ */
+# include "s390x_arch.h"
+
+/* Most modes of operation need km for partial block processing. */
+#  define S390X_aes_128_CAPABLE (OPENSSL_s390xcap_P.km[0] &  \
+                                S390X_CAPBIT(S390X_AES_128))
+#  define S390X_aes_192_CAPABLE (OPENSSL_s390xcap_P.km[0] &  \
+                                S390X_CAPBIT(S390X_AES_192))
+#  define S390X_aes_256_CAPABLE (OPENSSL_s390xcap_P.km[0] &  \
+                                S390X_CAPBIT(S390X_AES_256))
+
+/* Convert key size to function code: [16,24,32] -> [18,19,20]. */
+# define S390X_AES_FC(keylen)  (S390X_AES_128 + ((((keylen) << 3) - 128) >> 6))
+
+# define S390X_aes_128_gcm_CAPABLE (S390X_aes_128_CAPABLE &&        \
+                                    (OPENSSL_s390xcap_P.kma[0] &    \
+                                     S390X_CAPBIT(S390X_AES_128)))
+# define S390X_aes_192_gcm_CAPABLE (S390X_aes_192_CAPABLE &&        \
+                                    (OPENSSL_s390xcap_P.kma[0] &    \
+                                     S390X_CAPBIT(S390X_AES_192)))
+# define S390X_aes_256_gcm_CAPABLE (S390X_aes_256_CAPABLE &&        \
+                                    (OPENSSL_s390xcap_P.kma[0] &    \
+                                     S390X_CAPBIT(S390X_AES_256)))
+
+# define S390X_OUT_LEN    16
+# define S390X_IV_PAD_LEN 16
+/* iv + padding length for iv lengths != 12 */
+# define S390X_gcm_ivpadlen(i)  ((((i) + 15) >> 4 << 4) + S390X_IV_PAD_LEN)
+
+static int s390x_aes_gcm_init_key(PROV_AES_GCM_KEY *ctx,
+                                  const unsigned char *key, size_t keylen)
+{
+    ctx->key_set = 1;
+    memcpy(&ctx->plat.s390x.param.kma.k, key, keylen);
+    ctx->plat.s390x.fc = S390X_AES_FC(keylen);
+    if (!ctx->enc)
+        ctx->plat.s390x.fc |= S390X_DECRYPT;
+    return 1;
+}
+
+/*-
+ * Initialize context structure. Code is big-endian.
+ */
+static void s390x_aes_gcm_setiv(PROV_AES_GCM_KEY *ctx, const unsigned char *iv,
+                                size_t ivlen)
+{
+    S390X_KMA_PARAMS *kma = &ctx->plat.s390x.param.kma;
+
+    kma->t.g[0] = 0;
+    kma->t.g[1] = 0;
+    kma->tpcl = 0;
+    kma->taadl = 0;
+    ctx->plat.s390x.mreslen = 0;
+    ctx->plat.s390x.areslen = 0;
+    ctx->plat.s390x.kreslen = 0;
+
+    if (ivlen == AES_GCM_IV_DEFAULT_SIZE) {
+        memcpy(&kma->j0, iv, ivlen);
+        kma->j0.w[3] = 1;
+        kma->cv.w = 1;
+    } else {
+        unsigned long long ivbits = ivlen << 3;
+        size_t len = S390X_gcm_ivpadlen(ivlen);
+        unsigned char iv_zero_pad[AES_GCM_IV_MAX_SIZE + S390X_IV_PAD_LEN];
```

Comment:
S390X_gcm_ivpadlen(AES_GCM_IV_MAX_SIZE) is preferred, since it would still be correct, if AES_GCM_IV_MAX_SIZE was changed to a non-multiple of 16.